src/main/java/net/neoforged/neoforge/common/NeoForgeMod.java

```diff
@@ -591,34 +586,31 @@ public void serverStopping(ServerStoppingEvent evt) {
-        gen.addProvider(event.includeServer(), blockTags);
-        gen.addProvider(event.includeServer(), new NeoForgeItemTagsProvider(packOutput, lookupProvider, blockTags.contentsGetter(), existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeEntityTypeTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeFluidTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeEnchantmentTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeRecipeProvider(packOutput, lookupProvider));
-        gen.addProvider(event.includeServer(), new NeoForgeLootTableProvider(packOutput, lookupProvider));
-        gen.addProvider(event.includeServer(), new NeoForgeBiomeTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeStructureTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeDamageTypeTagsProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeServer(), new NeoForgeRegistryOrderReportProvider(packOutput));
-        gen.addProvider(event.includeServer(), new NeoForgeDataMapsProvider(packOutput, lookupProvider));
-
-        gen.addProvider(event.includeClient(), new NeoForgeSpriteSourceProvider(packOutput, lookupProvider, existingFileHelper));
-        gen.addProvider(event.includeClient(), new VanillaSoundDefinitionsProvider(packOutput, existingFileHelper));
-        gen.addProvider(event.includeClient(), new NeoForgeLanguageProvider(packOutput));
+
+        event.createProvider(includeServer, NeoForgeAdvancementProvider::new);
+        event.createBlockAndItemTags(NeoForgeBlockTagsProvider::new, NeoForgeItemTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeEntityTypeTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeFluidTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeEnchantmentTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeRecipeProvider::new);
+        event.createProvider(includeServer, NeoForgeLootTableProvider::new);
+        event.createProvider(includeServer, NeoForgeBiomeTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeStructureTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeDamageTypeTagsProvider::new);
+        event.createProvider(includeServer, NeoForgeRegistryOrderReportProvider::new);
+        event.createProvider(includeServer, NeoForgeDataMapsProvider::new);
+
+        event.createProvider(includeClient, NeoForgeSpriteSourceProvider::new);
+        event.createProvider(includeClient, VanillaSoundDefinitionsProvider::new);
+        event.createProvider(includeClient, NeoForgeLanguageProvider::new);
```

Comment:
Given these general usages, it might make more sense to have additional `createServerProvider` and `createClientProvider` methods instead of requiring the caller to pass `boolean run`. 

The current interface-based definition structure means this would lead to mass method overload though, so I'm not sure how clean that would be.

Reply:
> Given these general usages, it might make more sense to have additional `createServerProvider` and `createClientProvider` methods instead of requiring the caller to pass `boolean run`.

great idea, but it would tripple the method count (we still need one with parameter for dynamic scenarios or when using includeDev/includeReports). do we want this? I also talked about automatically detecting when to run what via some sort of marker in the provider class or pattern matching in addProvider.

> The current interface-based definition structure means this would lead to mass method overload though, so I'm not sure how clean that would be.

yes, i don't like that too but i couldn't come up with a better option. 
I did not want to use generic interfaces (like BiFunction) because I find them hard to read (`BiFunction<PackOutput, CompletableFuture<HolderLookup.Provider>, T>`). `DataProviderFromOutputLookupFileHelper<T>` is only slightly better though. I decided not to use `DataProvider.Factory<T>` for consistency but i could swap it to remove one interface.
As I mentioned before somewhere an alternative would be a depency injection system but I feel like that would be overkill. This would also fix cases where users reorder their parameters. I might consider it if there are other usecases for such a system. Alternativly we could use a library. (or does neo already has one?)

Reply:
@Shadows-of-Fire ?